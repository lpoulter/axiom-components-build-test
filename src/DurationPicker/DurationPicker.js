import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Duration } from 'luxon';
import omit from 'lodash.omit';
import Dropdown from '../Dropdown/Dropdown';
import DropdownContext from '../Dropdown/DropdownContext';
import DropdownMenu from '../Dropdown/DropdownMenu';
import DropdownMenuItem from '../Dropdown/DropdownMenuItem';
import DropdownSource from '../Dropdown/DropdownSource';
import DropdownTarget from '../Dropdown/DropdownTarget';
import Grid from '../Grid/Grid';
import GridCell from '../Grid/GridCell';
import TextInput from '../Form/TextInput';
import TextInputIcon from '../Form/TextInputIcon';

const validTimeUnits = ['seconds', 'minutes', 'hours', 'days', 'weeks', 'months', 'years'];

const getStateFromIsoDurationValue = (value) => {
  const duration = Duration.fromISO(value);

  if (!duration.isValid) {
    return {};
  }

  const selectedUnit = [].concat(validTimeUnits)
    .reverse()
    .find((timeUnit) => duration.values[timeUnit]);
  const selectedValue = duration.values[selectedUnit];

  return {
    selectedUnit,
    selectedValue,
    value,
  };
};

const getIsoDurationValueFromState = (state) => {
  const { selectedUnit, selectedValue } = state;

  if (!validTimeUnits.includes(selectedUnit) || !selectedValue) {
    return '';
  }

  const duration = Duration.fromObject({
    [selectedUnit]: selectedValue,
  });

  return duration.toISO();
};

const formatTimeUnit = (timeUnit) => {
  return `${timeUnit.slice(0, 1).toUpperCase()}${timeUnit.slice(1)}`;
};

export default class DurationPicker extends Component {
  static propTypes = {
    /** Excluded time-unit options */
    excludedOptions: PropTypes.arrayOf(PropTypes.oneOf(validTimeUnits)),
    /** Invoked when value or unit field has been blurred */
    onBlur: PropTypes.func,
    /** Invoked with the computed ISO duration value */
    onChange: PropTypes.func,
    /** Valid ISO duration value (see https://en.wikipedia.org/wiki/ISO_8601#Durations) [DISABLED] */
    value: PropTypes.string,
  };

  static defaultProps = {
    excludedOptions: ['seconds'],
    onBlur: () => {},
    onChange: () => {},
    value: 'P7D',
  };

  constructor(props) {
    super(props);

    this.onBlur = this.onBlur.bind(this);
    this.onChangeValue = this.onChangeValue.bind(this);
    this.onChangeUnit = this.onChangeUnit.bind(this);
    this.onChange = this.onChange.bind(this);

    this.state = getStateFromIsoDurationValue(props.value);
  }

  onBlur() {
    this.props.onBlur();
  }

  onChangeValue(event) {
    const inputValue = parseInt(event.target.value, 10);
    const selectedValue = Number.isFinite(inputValue) && inputValue >= 1 ? inputValue : '';
    const durationValue = getIsoDurationValueFromState({
      selectedUnit: this.state.selectedUnit,
      selectedValue,
    });

    this.setState({
      selectedValue,
      value: durationValue,
    }, this.onChange);
  }

  onChangeUnit(selectedUnit) {
    const durationValue = getIsoDurationValueFromState({
      selectedUnit,
      selectedValue: this.state.selectedValue,
    });

    this.setState({
      selectedUnit,
      value: durationValue,
    }, this.onChange);
  }

  onChange() {
    this.props.onChange(this.state.value);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.value !== this.props.value) {
      this.setState(getStateFromIsoDurationValue(this.props.value), this.onChange);
    }
  }

  render() {
    const { excludedOptions } = this.props;
    const childProps = omit(this.props, Object.keys(DurationPicker.propTypes));
    const filteredTimeUnits = validTimeUnits.filter((timeUnit) => !excludedOptions.includes(timeUnit));

    const {
      selectedUnit = filteredTimeUnits[0],
      selectedValue = '',
    } = this.state;

    return (
      <Grid
          fit={ true }
          gutters="tiny"
          responsive={ false }
          space="x0">
        <GridCell>
          <TextInput
              { ...childProps }
              onBlur={ this.onBlur }
              onChange={ this.onChangeValue }
              type="number"
              value={ (`${selectedValue}`) } />
        </GridCell>
        <GridCell>
          <Dropdown
              onRequestClose={ this.onBlur }>
            <DropdownTarget>
              <TextInput
                  { ...childProps }
                  isTarget
                  readOnly
                  value={ formatTimeUnit(selectedUnit) }>
                <TextInputIcon
                    name="chevron-down" />
              </TextInput>
            </DropdownTarget>
            <DropdownSource>
              <DropdownContext>
                <DropdownMenu>
                  {
                    filteredTimeUnits.map((timeUnit) => {
                      return (
                        <DropdownMenuItem
                            key={ timeUnit }
                            onClick={ () => this.onChangeUnit(timeUnit) }
                            selected={ timeUnit === selectedUnit }>
                          { formatTimeUnit(timeUnit) }
                        </DropdownMenuItem>
                      );
                    })
                  }
                </DropdownMenu>
              </DropdownContext>
            </DropdownSource>
          </Dropdown>
        </GridCell>
      </Grid>
    );
  }
}
