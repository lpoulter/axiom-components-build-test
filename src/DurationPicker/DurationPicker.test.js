import React from 'react';
import { shallow } from 'enzyme';
import DurationPicker from './DurationPicker';
import toJson from 'enzyme-to-json';

const render = (props = {}, opts = {}) => shallow(<DurationPicker { ...props } />, opts);

describe('DurationPicker', () => {
  describe('Snapshots', () => {
    it('renders with defaultProps', () => {
      expect(toJson(render({}))).toMatchSnapshot();
    });
  });
});
