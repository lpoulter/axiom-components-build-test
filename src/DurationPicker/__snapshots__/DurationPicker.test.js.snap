// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationPicker Snapshots renders with defaultProps 1`] = `
<Grid
  direction="row"
  fit={true}
  gutters="tiny"
  responsive={false}
  space="x0"
  wrap={true}
>
  <GridCell>
    <TextInput
      inlineLabel={false}
      onBlur={[Function]}
      onChange={[Function]}
      size="medium"
      space="x4"
      type="number"
      value="7"
    />
  </GridCell>
  <GridCell>
    <Dropdown
      enabled={true}
      onRequestClose={[Function]}
      position="bottom"
      showArrow={false}
    >
      <DropdownTarget>
        <TextInput
          inlineLabel={false}
          isTarget={true}
          readOnly={true}
          size="medium"
          space="x4"
          type="text"
          value="Days"
        >
          <TextInputIcon
            align="right"
            name="chevron-down"
          />
        </TextInput>
      </DropdownTarget>
      <DropdownSource>
        <DropdownContext
          maxHeight="30rem"
          position="top"
          width="14.5rem"
        >
          <DropdownMenu>
            <DropdownMenuItem
              key="minutes"
              onClick={[Function]}
              selected={false}
            >
              Minutes
            </DropdownMenuItem>
            <DropdownMenuItem
              key="hours"
              onClick={[Function]}
              selected={false}
            >
              Hours
            </DropdownMenuItem>
            <DropdownMenuItem
              key="days"
              onClick={[Function]}
              selected={true}
            >
              Days
            </DropdownMenuItem>
            <DropdownMenuItem
              key="weeks"
              onClick={[Function]}
              selected={false}
            >
              Weeks
            </DropdownMenuItem>
            <DropdownMenuItem
              key="months"
              onClick={[Function]}
              selected={false}
            >
              Months
            </DropdownMenuItem>
            <DropdownMenuItem
              key="years"
              onClick={[Function]}
              selected={false}
            >
              Years
            </DropdownMenuItem>
          </DropdownMenu>
        </DropdownContext>
      </DropdownSource>
    </Dropdown>
  </GridCell>
</Grid>
`;
